src/main/java/com/google/sps/data/algorithm/TspSolver.java

```diff
@@ -0,0 +1,414 @@
+      timeMatrix[i][i] = 0;
+    }
+
+    // fill (i, j) and (j, i) indices with travel time from i to j
+    for (int i = 0; i < allPois.size(); i++) {
+      for (int j = i; j < allPois.size(); j++) {
+        int travelTime = getTravelTimeMins(allPois.get(i), allPois.get(j));
+        timeMatrix[i][j] = travelTime;
+        timeMatrix[j][i] = travelTime;
+      }
+    }
+  }
+
+  /**
+   * Populate placeIdToInt and call setOpenHours for each placeId
+   *
+   * @param center central location
+   * @param pois list of pois to visit
+   */ 
+  private void populateIntMapAndOpenHours(String center, 
+      List<String> pois) throws IOException{
+    this.openHours = new HashMap<>();
+    this.intToPlaceId = new HashMap<>();
+    this.intToPlaceId.put(0, center);
+    setOpenHours(0, center);
+
+    for (int i = 0; i < pois.size(); i++) {
+      this.intToPlaceId.put(i+1, pois.get(i));
+      setOpenHours(i+1, pois.get(i));
+    }
+  }
+
+  /**
+   * Sets the openHours for each placeId (as int), as an OpeningHours.Period
+   * If can't find opening hours set to be null (represents open all day)
+   * 
+   * @param placeId placeId to search openHours for
+   * @param index index to put the openHours in the HashMap
+   */
+  private void setOpenHours(int index, String placeId) throws IOException {
+    PlaceDetails placeDetails = TripServlet.getPlaceDetailsFromPlaceId(this.context, placeId);
+
+    if (placeDetails.openingHours == null) { 
+      this.openHours.put(index, null);    
+    } else if (placeDetails.openingHours.periods.length != 7) {
+      checkDayIsPresentInHours(placeDetails.openingHours, index);
+    } else {
+      this.openHours.put(index, placeDetails.openingHours.periods[this.intOfWeek]);
+    } 
+  }
+
+  /**
+   * If day in openingHours array is the same as the day searched for 
+   * check if open and close time are not null. If no hours found put null
+   * into openHours (represents open all day)
+   * 
+   * @param openingHours openHours for placeId on desired day
+   * @param index index to put result in OpenHours Map
+   */
+  private void checkDayIsPresentInHours(OpeningHours openingHours, int index) {
+      int adjustedWeekIndex;
+      // periods is numbered Sunday - Saturday, WeekdayText is Monday - Sunday
+      if (intOfWeek == 0) {
+        adjustedWeekIndex = 6;
+      } else {
+        adjustedWeekIndex = intOfWeek - 1;
+      }
+      // string representation of day searched
+      String desiredDayString = WEEKDAYS[adjustedWeekIndex];
+      for (int i = 0; i < openingHours.periods.length; i++) {
+        // get the day that the ith index in opening hours corresponds to
+        String day = openingHours.weekdayText[i].split(":")[0];
+
+        // check neither open nor closing time are null
+        boolean open = openingHours.periods[i].open != null;
+        boolean close = openingHours.periods[i].close != null;
+
+        if (day.equals(desiredDayString) && open && close) {
```

Comment:
nit: it seems like day.equals(desiredDayString) is always going to be true? If so you might not need the WEEKDAYS constant.

Reply:
It won't always return true because I'm checking that the location is open for the specific day I'm looking for. So if I'm looking for open hours of "Tuesday", `desiredDayString` will be "Tuesday" but if `i=0` then `day` will be "Monday" and the if statement will be entered and put in wrong hours (Tuesday may not have any open hours).